File: 503-next-greater-element-ii/503-next-greater-element-ii.java
class Solution {
    
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Stack<Integer> stack = new Stack<>();
        for(int i = 0; i < n * 2; i++){
            int num = nums[i % n];
            while(!stack.isEmpty() && num > nums[stack.peek()]){
                result[stack.pop()] = num;
            }
            if(i < n){
                stack.push(i);
            }
        }
        return result;
    }

}

File: 503-next-greater-element-ii/503-next-greater-element-ii.cpp
class Solution {

public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        
        int size1 = nums.size() , i , max_num = INT_MIN , j;
        vector<int> v;
        for( i = 0 ; i < size1 ; i++ )
        {
            if( max_num < nums[i] )
                max_num = nums[i];
        }
        for( i = 0 ; i < size1 ; i++ )
        {
            for( j = i + 1 ; j < size1 ; j++ )
            {
                if( nums[j] > nums[i] )
                {
                    v.push_back( nums[j] );
                    break;
                }
            }
            if( j == size1 )
            {
                for( j = 0 ; j < i ; j++ )
                {
                    if( nums[j] > nums[i] )
                    {
                        v.push_back( nums[j] );
                        break;
                    }
                }
                if( j == i )
                    v.push_back(-1);
            }
        }
        return v;
    }
};